### PR TITLE
r/digitalocean_database_cluster: Adding new resource to manage database clusters

### PR DESCRIPTION
**Pull request by slapula**
```
$ make testacc TESTARGS='-run=TestAccDigitalOceanDatabaseCluster_'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test $(go list ./... |grep -v 'vendor') -v -run=TestAccDigitalOceanDatabaseCluster_ -timeout 120m
?   	github.com/terraform-providers/terraform-provider-digitalocean	[no test files]
=== RUN   TestAccDigitalOceanDatabaseCluster_Basic
--- PASS: TestAccDigitalOceanDatabaseCluster_Basic (269.01s)
=== RUN   TestAccDigitalOceanDatabaseCluster_WithUpdate
--- PASS: TestAccDigitalOceanDatabaseCluster_WithUpdate (475.48s)
PASS
ok  	github.com/terraform-providers/terraform-provider-digitalocean/digitalocean	744.505s
```

### CODE REVIEW

File: website/digitalocean.erb [hunk 1]
```diff
@@ -16,6 +16,9 @@
             <li<%= sidebar_current("docs-do-datasource-certificate") %>>
               <a href="/docs/providers/do/d/certificate.html">digitalocean_certificate</a>
             </li>
+            <li<%= sidebar_current("docs-do-resource-database-cluster") %>>
```
[the remark this answers is not in the document]
Comment (slapula) [reply website/digitalocean.erb]: Oops.  Fixed.

File: digitalocean/resource_digitalocean_database_cluster.go [hunk 2]
```diff
@@ -0,0 +1,196 @@
...
+			"size": {
+				Type:         schema.TypeString,
+				Required:     true,
+				ValidateFunc: validation.NoZeroValues,
+			},
+
+			"region": {
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
```
[the remark this answers is not in the document]
Comment (slapula) [reply digitalocean/resource_digitalocean_database_cluster.go]: Now that I looked at it, this was pretty easy to implement.  I went ahead and added it to the PR and included the corresponding acceptance test:
```
$ make testacc TESTARGS='-run=TestAccDigitalOceanDatabaseCluster_WithMigration'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test $(go list ./... |grep -v 'vendor') -v -run=TestAccDigitalOceanDatabaseCluster_WithMigration -timeout 120m
?   	github.com/terraform-providers/terraform-provider-digitalocean	[no test files]
=== RUN   TestAccDigitalOceanDatabaseCluster_WithMigration
--- PASS: TestAccDigitalOceanDatabaseCluster_WithMigration (594.58s)
PASS
ok  	github.com/terraform-providers/terraform-provider-digitalocean/digitalocean	594.590s
```

**Comment by slapula** [comment]
Added the ability to manage maintenance windows as well:
```
$ make testacc TESTARGS='-run=TestAccDigitalOceanDatabaseCluster_WithMaintWindow'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test $(go list ./... |grep -v 'vendor') -v -run=TestAccDigitalOceanDatabaseCluster_WithMaintWindow -timeout 120m
?   	github.com/terraform-providers/terraform-provider-digitalocean	[no test files]
=== RUN   TestAccDigitalOceanDatabaseCluster_WithMaintWindow
--- PASS: TestAccDigitalOceanDatabaseCluster_WithMaintWindow (296.92s)
PASS
ok  	github.com/terraform-providers/terraform-provider-digitalocean/digitalocean	296.933s
```

File: digitalocean/resource_digitalocean_database_cluster.go [hunk 3]
```diff
@@ -0,0 +1,336 @@
...
+	}
+
+	d.Set("name", database.Name)
+	d.Set("engine", database.EngineSlug)
+	d.Set("version", database.VersionSlug)
+	d.Set("size", database.SizeSlug)
+	d.Set("region", database.RegionSlug)
+	d.Set("node_count", database.NumNodes)
+
+	if err := d.Set("maintenance_window", flattenMaintWindowOpts(*database.MaintenanceWindow)); err != nil {
```
Comment (andrewsomething) [inline digitalocean/resource_digitalocean_database_cluster.go]: This is generating a diff when `terraform apply` is run twice in a row if a maintenance window is not defined leading to the other tests failing:

```
=== RUN   TestAccDigitalOceanDatabaseCluster_Basic
--- FAIL: TestAccDigitalOceanDatabaseCluster_Basic (367.44s)
    testing.go:538: Step 0 error: After applying this step, the plan was not empty:
        
        DIFF:
        
        UPDATE: digitalocean_database_cluster.foobar
          maintenance_window.#:      "1" => "0"
          maintenance_window.0.day:  "saturday" => ""
          maintenance_window.0.hour: "06:44:04" => ""
```

We should probably only set to state if there is a user defined window.

**Comment by slapula** [comment]
Oh, interesting. Ok, I'll take a look at this later tonight.

File: digitalocean/resource_digitalocean_database_cluster.go (hunk 3 above)
Comment (slapula) [reply digitalocean/resource_digitalocean_database_cluster.go]: I think I fixed this with my latest commit.  All tests are passing for me now:
```
$ make testacc TESTARGS='-run=TestAccDigitalOceanDatabaseCluster_'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test $(go list ./... |grep -v 'vendor') -v -run=TestAccDigitalOceanDatabaseCluster_ -timeout 120m
?   	github.com/terraform-providers/terraform-provider-digitalocean	[no test files]
=== RUN   TestAccDigitalOceanDatabaseCluster_Basic
--- PASS: TestAccDigitalOceanDatabaseCluster_Basic (314.54s)
=== RUN   TestAccDigitalOceanDatabaseCluster_WithUpdate
--- PASS: TestAccDigitalOceanDatabaseCluster_WithUpdate (593.32s)
=== RUN   TestAccDigitalOceanDatabaseCluster_WithMigration
--- PASS: TestAccDigitalOceanDatabaseCluster_WithMigration (653.17s)
=== RUN   TestAccDigitalOceanDatabaseCluster_WithMaintWindow
--- PASS: TestAccDigitalOceanDatabaseCluster_WithMaintWindow (240.31s)
PASS
ok  	github.com/terraform-providers/terraform-provider-digitalocean/digitalocean	1801.345s
```

**Comment by ehamberg** [comment]
Thank you so much for doing this, @slapula!

I've been playing around with this and ended up having to recreate a cluster. Applying that change gave me the following error:

    * digitalocean_database_cluster.ehtest-staging: Error creating DatabaseCluster:
      POST https://api.digitalocean.com/v2/databases: 422
      (request "e48dd728-1951-489e-a690-d9703ac5c876") cluster name is not available

I could kind of guess what was happening behind the scenes (basically that the destroy hadn't finished yet when it was attempting to create a new cluster), and – as expected –  retrying a little while later worked.

Not sure if this is something Terraform providers usually deal with as I'm new to Terraform, but wanted to mention this just in case. (Any way, this is not a blocker.)

**Comment by slapula** [comment]
@ehamberg this is something I noticed as well but if I recall correctly there is no 'deleting' status to craft a waiter around (well not easily in my case I'm no golang ninja). I'm going to play around with it and see if I can resolve this.

**Comment by andrewsomething** [comment]
@ehamberg Thanks for the testing and feedback! I've tried to reproduce this a few times without success. Talking to folks working on DBaaS at DigitalOcean, the name should be released immediately. They're going to take a closer look to see if there is anything that can be improved on that end. On the Terraform side, AFAIK there isn't a mechanism to determine if a resource is being created for the first time or being recreated. So if we need to protect against this, we'd probably need to catch that error message and retry after a small sleep. Of course it can also be a valid error, so we'd probably only want to do that once. So let's follow up on this latter if needed.



:pray: Thanks again for all your work on this @slapula!

**Comment by eddiezane** [comment]
@slapula thanks a ton for this PR!

Would you please shoot me an email when you get a chance? My GH username at digitalocean dot com.

**Pull request merged**